In [1]:
import pandas as pd
import numpy as np

In [2]:
!ls

1. Data Cleaning (Basic).ipynb
2. Data Analysis and Cleaning (Advance).ipynb
3. Data Prep.ipynb
4. Model_Training.ipynb
5. Prediction Pipeline.ipynb
Applications
CNN_Face_recognition.ipynb
Data_Prepared.csv
Desktop
Documents
Downloads
Library
Model
Movies
Music
ND_DS_LSTM.ipynb
ORL_faces.npz
Pictures
Public
anaconda3
clean_data.csv
cleaned_test_booking_data.csv
data_analysis_ride_request.html
frs_4_Predictions copy.ipynb
model_VAR.pkl
preprocessed_1.csv
preprocessed_2.csv
preprocessed_3.csv
raw_data.csv
templates
water-consumption-in-the-new-york-city.csv


#### Reading DataSet

In [3]:
df = pd.read_csv('raw_data.csv', low_memory = False, compression='gzip')
len(df)

8381556

In [4]:
df.head()

ts number   pick_lat   pick_lng   drop_lat   drop_lng
0  2020-03-26 07:07:17  14626  12.313621  76.658195  12.287301  76.602280
1  2020-03-26 07:32:27  85490  12.943947  77.560745  12.954014  77.543770
2  2020-03-26 07:36:44  05408  12.899603  77.587300  12.934780  77.569950
3  2020-03-26 07:38:00  58940  12.918229  77.607544  12.968971  77.636375
4  2020-03-26 07:39:29  05408  12.899490  77.587270  12.934780  77.569950

### A Customer_ID `number` at a particular timestamp can only have one entry
### Removing Duplicate Entries ['ts','number']

In [5]:
df[df.duplicated(subset=['ts','number'],keep=False)]

ts number   pick_lat   pick_lng   drop_lat  \
235      2020-03-26 18:10:35  16795  12.967236  77.641594  13.014504   
236      2020-03-26 18:10:35  16795  12.967236  77.641594  13.014504   
407      2020-03-26 21:35:50  65856  12.917173  77.586400  12.913940   
408      2020-03-26 21:35:50  65856  12.917173  77.586400  12.913940   
443      2020-03-26 23:26:29  27554  12.933715  77.619300  12.938208   
...                      ...    ...        ...        ...        ...   
8381231  2021-03-26 22:23:12  61636  12.975229  77.620370  13.017285   
8381245  2021-03-26 22:25:13  61636  12.975229  77.620370  13.017285   
8381246  2021-03-26 22:25:13  61636  12.975229  77.620370  13.017285   
8381248  2021-03-26 22:25:27  61636  12.975229  77.620370  13.017285   
8381249  2021-03-26 22:25:27  61636  12.975229  77.620370  13.017285   

          drop_lng  
235      77.650856  
236      77.650856  
407      77.685280  
408      77.685280  
443      77.587520  
...            ...  
8381231  77.618200  
8381245  77.618200  
8381246  77.618200  
8381248  77.618200  
8381249  77.618200  

[113540 rows x 6 columns]

### There are 113540 Duplicate Entries
#### We have 8315498 Unique timestamp, customer_id rows. 

In [6]:
## Keeping first occurence
df.drop_duplicates(subset=['ts','number'], inplace = True, keep = 'last')

df.reset_index(inplace = True, drop = True)

In [7]:
# Info of Dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8315498 entries, 0 to 8315497
Data columns (total 6 columns):
 #   Column    Dtype  
---  ------    -----  
 0   ts        object 
 1   number    object 
 2   pick_lat  float64
 3   pick_lng  float64
 4   drop_lat  float64
 5   drop_lng  float64
dtypes: float64(4), object(2)
memory usage: 380.7+ MB


In [8]:
#Count missing values
np.count_nonzero(df.isnull().values)

0

In [9]:
df['number'] = pd.to_numeric(df['number'], errors = 'coerce')

#Count missing values
np.count_nonzero(df.isnull().values)

116

#### There are 116 NaN rows, dropping NaN rows.

In [10]:
df.dropna(inplace = True)
len(df)

8315382

In [11]:
df['number'] = pd.to_numeric(df['number'], errors = 'coerce', downcast='integer')
df['ts'] = pd.to_datetime(df['ts'])

In [12]:
# Info of Dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8315382 entries, 0 to 8315497
Data columns (total 6 columns):
 #   Column    Dtype         
---  ------    -----         
 0   ts        datetime64[ns]
 1   number    int32         
 2   pick_lat  float64       
 3   pick_lng  float64       
 4   drop_lat  float64       
 5   drop_lng  float64       
dtypes: datetime64[ns](1), float64(4), int32(1)
memory usage: 412.4 MB


### Breaking Time to Features

In [13]:
df['hour'] = df['ts'].dt.hour
df['mins'] = df['ts'].dt.minute
df['day'] = df['ts'].dt.day
df['month'] = df['ts'].dt.month
df['year'] = df['ts'].dt.year
df['dayofweek'] = df['ts'].dt.dayofweek

In [14]:
df.to_csv('preprocessed_1.csv',index = False, compression = 'gzip')

In [18]:
print(df.shape)

(8315382, 12)
